internal/expr/parser.go

```diff
@@ -129,25 +129,81 @@ func (p *Parser) Parse(input string) (expr *ParsedExpr, err error) {
 			continue
 		}
 
-		if sp, ok, err := p.parseStringLiteral(); err != nil {
-			return nil, err
-		} else if ok {
-			p.add(sp)
-			continue
-		}
-
 		if p.pos == len(p.input) {
 			break
 		}
 
-		// If nothing above can be parsed we advance the parser.
-		p.pos++
+		// If nothing above can be parsed we advance the parser to the next
+		// character of interest.
+		if err := p.advance(); err != nil {
+			return nil, err
+		}
 	}
 	// Add any remaining unparsed string input to the parser.
 	p.add(nil)
 	return &ParsedExpr{p.parts}, nil
 }
 
+// advance increments p.pos until it reaches content that might preceed a token
+// we want to parse.
+func (p *Parser) advance() error {
+
+	// The byte following these bytes might be the start of an expression.
+	delimiterBytes := map[byte]bool{
+		' ':  true,
+		'\t': true,
+		'\n': true,
+		'\r': true,
+		')':  true,
+		'(':  true,
+		';':  true,
+		'=':  true,
+		',':  true,
+	}
+
+	for p.pos < len(p.input) {
+		if ok, err := p.skipStringLiteral(); err != nil {
+			return err
+		} else if ok {
+			continue
+		}
+		p.pos++
+		if p.pos < len(p.input) && delimiterBytes[p.input[p.pos]] {
```

Comment:
Maybe we don't need `p.pos < len(p.input)` here. Because when `p.pos == len(p.input)`, the loop will break immediately after anyway :)
This made me think: when `p.pos == len(p.input)`, are any of these delimiter characters supposed to be not there? In other words, is it an error to find any of these characters at the end of the input? (e.g. '=') I may be missing the context here of course :)

Reply:
We have it here because of the `p.pos++` on the line before.

Generally we dont worry about any characters we don't recognise as our own. If the user has put an `=` at the end of the SQL then that's their problem.

Reply:
If I'm getting it correctly, `skipStringLiteral()` returns `false, nil` when the byte at the current position is not a single or double quote. In such a case, it doesn't skip that byte. I believe, after the `if-else` block here, it's this same case where `ok == false` and `err==nil` from `p.skipStringLiteral()`. At that point, perhaps `p.pos++` shouldn't be here, because we then end up not checking the character at the current position (which wasn't skipped by `skipStringLiteral()` at this point. So it feels like perhaps `p.pos++` needs to be after we check the byte at the current position for the set of delimiters listed, i.e. after this `if-block` (and in that case, `p.pos < len(p.input)` in the `if-condition` seems to be redundant, because the `for` loop condition has already checked it.

Reply:
Fixed with a small shuffle around.